### PR TITLE
Create/Edit Channel layout

### PR DESCRIPTION
**Pull request by mikkio-j**
#335

### CODE REVIEW

File: src/shared/components/TextArea/TextArea.style.ts
Comment (drillprop) [inline src/shared/components/TextArea/TextArea.style.ts]: What if we want to use a higher text-area? I know that is not the scope of this issue but maybe we can use the text-area `rows` attribute here

File: src/shared/components/TextArea/TextArea.style.ts
Comment (drillprop) [inline src/shared/components/TextArea/TextArea.style.ts]: Do we really need that? I think we could set some default constant value for the line-height

File: src/views/PlaygroundView.tsx
Comment (drillprop) [inline src/views/PlaygroundView.tsx]: Not sure why you disable this. Everything seems to be fine here, right?

File: src/views/Playgrounds/PlaygroundChannelCreate.tsx
Comment (drillprop) [inline src/views/Playgrounds/PlaygroundChannelCreate.tsx]:
```suggestion
const PlaygroundChannelCreate = () => {
```

File: src/views/Playgrounds/PlaygroundChannelCreate.tsx
Comment (drillprop) [inline src/views/Playgrounds/PlaygroundChannelCreate.tsx]:
```suggestion
export default PlaygroundChannelCreate
```

File: src/views/Playgrounds/PlaygroundChannelEdit.tsx
Comment (drillprop) [inline src/views/Playgrounds/PlaygroundChannelEdit.tsx]: Small thing. If you pick values inside select inputs and then click cancel - selected values will not update. I think it might be an issue with the select input itself.

**Review by drillprop — COMMENTED** [review]
Again, besides the mirage thing, looks solid. Well done

**Comment by eldiegod** [comment]
Gotta solve some conflicts over here

**Review by eldiegod — CHANGES_REQUESTED** [review]
LGTM overall just a visual thing I noticed; could we add a transition to when we hover the video cover it feels a bit rough?

**Comment by drillprop** [comment]
Also, it would be nice to hide tooltips if the user click inputs or the text-area

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]:
```suggestion
            description="Channel language is the main language of the content you publish on your channel"
```

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]:
```suggestion
            description="Publicness dictates whether your channel will be displayed in users' feeds"
```

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]: This should be `Create a Channel` when not editing

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]: I think we will need to save the blob as well, URL may not be enough to upload the file since we would need to read them again to get the data

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]: We will need to populate this with some more languages. Also, the key should probably be the ISO 639-1 identifier of the language as that's what's in the schema. I suggest to find some list of the most popular languages and put the first 20 or so here. We can revisit later

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]:
```suggestion
  { name: 'Public (Channel will appear in feeds)', value: 'public' },
```

File: src/views/Playgrounds/PlaygroundChannelEditCreate.tsx
Comment (kdembler) [inline src/views/Playgrounds/PlaygroundChannelEditCreate.tsx]: Better value here would probably be `Unlisted` since the channel will be always public (it's in the blockchain after all). Also the value should be `true/false`. The corresponding field is `isPublic`

**Review by kdembler — CHANGES_REQUESTED** [review]
Will take a look at the code tomorrow, for now just looking at the app. Some observations:
1) I think we should actually put this at a proper route already, to see how it looks as a separate view. We can wait till #359 is merged to get the basic `/studio` routing. My guess is the route will look like `/studio/channel/{id}` or `/studio/channel/new`. Based on that we will know whether it's edit or a new channel
2) Are you rebased on latest button changes? It seems the `Cancel` and `Publish` buttons have different heights
3) The `Publish` button doesn't have a `pointer` cursor
4) We should pre-populate value for both selects. In case of editing a channel we should obviously display the channel's values. When a new channel is created we can use defaults: `Public` and `English`
5) The bottom action bar should only be visible when:
    a) it's a new channel
    b) there were some changes made

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: I think we will need to be a tad more liberal here, I suggest maybe like 40 characters max

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx:193]: This will make it required, right? I think we can make the description optional

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: We certainly shouldn't rely on anything from playground outside of it. If it has something we need, let's elevate it higher

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: This looks like a much better list! I have two suggestions though:
1. Let's sort it alphabetically so it's easier to find your desired language (we can keep English at the top as the default option)
2. Maybe let's elevate this somewhere above to `config/` or elsewhere so this file is easier to read (we shouldn't use `SelectedItem` type there though)

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: I don't think it makes much sense to do a call fetching channel with `{id: ''}` - this will always be `null`. We should skip this call when dealing with a new channel

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: This will get slightly more complicated once we do #269 - then we won't necessarily have a url for an asset. Or resolving that URL can be async, depending on what we do. I guess we can leave it for now, just FYI

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: Is this blocked atm?

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: Do we need a separate `useEffect` for this? Cannot those be set at the same time as `setValue` calls are made?

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx:78]: Let's move those above `useEffect`s, we use some state inside them

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx:242]: Would it be a lot of work to show this only if there are some changes to be submitted? For example, if I remove the last character of channel name and input it a gain, the form will be dirty even though there are no changes to actually publish

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: This was fine for the playground view, but the final thing shouldn't do this. The question then is, what should happen when we click that. Truth be told, I'm not sure, maybe let's just keep it empty for now and I ask designers

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx:144]: Why is this needed?

File: src/shared/components/ActionBar/ActionBar.style.ts
Comment (kdembler) [inline src/shared/components/ActionBar/ActionBar.style.ts]: We're using a Button component for those buttons right? Shouldn't they have that applied already?

File: src/shared/components/ActionBar/ActionBar.style.ts
Comment (kdembler) [inline src/shared/components/ActionBar/ActionBar.style.ts]:
```suggestion
  > *:not(:last-child) {
```

This is a minor change but:
- by using `> *` we make sure the margin works for whatever is here, for example if we want to make the cancel button a link, this will work, `button` will not catch that
- with `nth-of-type(1)` we will add margin only to one of the buttons, if we add more of them, we'd need to update the styles. with `:not(:last-child)` we can target all items except the last one which is what I believe we want to do

Let me know if something here doesn't make sense

File: src/shared/components/HeaderTextField/HeaderTextField.style.tsx
Comment (kdembler) [inline src/shared/components/HeaderTextField/HeaderTextField.style.tsx]: Could you explain whats going on here? Seems there's a lot of arbitrary values - `150px`, `200px`, `sizes(2.5)`

File: src/shared/components/TextArea/TextArea.style.ts
Comment (kdembler) [inline src/shared/components/TextArea/TextArea.style.ts]: Per design text here is `Body 1`, so we should use values from the theme that correspond to it
```suggestion
  font-size: ${typography.sizes.body1};
  line-height: ${typography.lineHeights.body1};
```

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx]: I'm a bit unsure about this and the `margin-left` below but I think those may be affected by the max view width we discussed in the chat so let's see how it goes

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx]: I think this may be a bit too much on mobile

**Review by kdembler — CHANGES_REQUESTED** [review]
It's great to see all these coming together. Did a deeper dive this time and found some things we can improve, nothing major though!

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (mikkio-j) [reply src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx:144]: Hook form captures click and triggers submission

File: src/shared/components/HeaderTextField/HeaderTextField.style.tsx
Comment (mikkio-j) [reply src/shared/components/HeaderTextField/HeaderTextField.style.tsx]: `HeaderTextField` changes it's width depends on its input's length, so it was little bit tricky to style for all devices and different input's length. Changed that now to more cleaner option, I believe

File: src/shared/components/ActionBar/ActionBar.style.ts
Comment (eldiegod) [inline src/shared/components/ActionBar/ActionBar.style.ts]: This will have to be removed after merging #385 that handles that for all studio routes

File: src/shared/components/ActionBar/ActionBar.style.ts
Comment (eldiegod) [inline src/shared/components/ActionBar/ActionBar.style.ts]: same as above

File: src/views/LayoutWithRouting.tsx
Comment (eldiegod) [inline src/views/LayoutWithRouting.tsx]: Same same as above #385

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx
Comment (eldiegod) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx]: Not sure if you would still need this one but this variable will still exists after #385 is merged

**Review by eldiegod — APPROVED** [review]
Looks good to me; added some comments on some things that we will need to be removed as we merge the `Layout for studio routes` PR.

As a side note, the channel cover img hovering effect turning darker could use a transition

File: src/api/queries/__generated__/channels.generated.tsx
Comment (kdembler) [inline src/api/queries/__generated__/channels.generated.tsx]: This breaks the consumer views in mocked version since we're not returning this info

**Review by kdembler — COMMENTED** [review]
It seems if I have my cursor below the description text area, the `click anywhere to edit cover image` is still displayed. I think we will need to restrict the hover area for that since it produces not ideal experience:

https://user-images.githubusercontent.com/12646744/111434877-5de57280-8700-11eb-865b-a0fecdc07a1a.mov

File: src/shared/components/HeaderTextField/HeaderTextField.style.tsx
Comment (kdembler) [inline src/shared/components/HeaderTextField/HeaderTextField.style.tsx]: I think we're trying to restrict width of this component too much. Take a look what happens across screen widths:

Screenshot 2021-03-17 at 15 45 25

Screenshot 2021-03-17 at 15 47 59

I feel this could be wider to offer better support to longer channel names. Look into this please

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx]: I don't think this should be `40px`. If you look at the design, `40px` is indeed spacing between end of one select to the header of the another but our components also have a reserved space for the helper text which will make this spacing actually bigger. I think the value used in the component is correct.

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.style.tsx]: I believe these fields should be consistent across views so if some changes are needed to those, we should just put the styles into the component so each view doesn't have to style those separately

File: src/shared/components/ActionBar/ActionBar.style.ts
Comment (kdembler) [inline src/shared/components/ActionBar/ActionBar.style.ts:21]: Not related to this component specifically, but I think we could adjust the breakpoints here a bit:
Screenshot 2021-03-17 at 15 59 11

It seems there's enough space to display the `Every change...` next to the fee instead of below it

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: Why do we need those? We already call those setters in `handleAvatarConfirm` and `handleCoverConfirm`

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [inline src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: I think this can be just `isPublic: isPublic`?

**Review by kdembler — COMMENTED** [review]
Some final tweaks

File: src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx
Comment (kdembler) [reply src/views/studio/CreateEditChannelView/CreateEditChannelView.tsx]: This hook will be called a lot of times

**Review by kdembler — APPROVED** [review]
Good job!

**Pull request merged**
